### DISCUSSION TITLE
Go API Internalization

### DISCUSSION THREAD

**Discussion by tychoish** [tendermint/tendermint#6423] [Q&A]
From my perspective, and our internal conversations there are a few major schools of thoughts with how this will go and I just wanted to outline some of the big questions that I think will help shape what we actually want to do:

- How do we expect that users of tendermint, as a library, will construct tendermint objects? 

- What are the use cases for tendermint as a library?

- To what extent do we want to allow users to customize specific component services via dependency injection?

- Should components of tendermint (e.g. mempool, blockstores, etc.) be constructable or replaceable by users of tendermint as a library?

For gratuitous background, for an API to be public, it needs to have the following properties: 

- it needs to be **public/exported** (e.g. not in an `internal` directory,) have a type name that begins in a capital letter. For public struct types, all exported fields must also refer to public types. 

- **constructable**, it must be possible to construct exported types without depending on private/internal/non-exported either using literal construction or constructor functions. 

- all of its **dependencies must also be public and constructable**, including all transitive/interface dependencies (function arguments and return values).

The requirements for constructability and similar dependencies in practice mean that larger sections of the type graph need to move together. Paring constructors/factory functions that return public interfaces to wrap private implementations is often a useful strategy for minimizing the exported surface of a package without and reducing hard dependencies between packages.

The approaches and decisions, then are: 

- should we move `node`, `p2p`, `rpc` and all of the reactors into an `internal` package, leaving `cmd`, `config`, `libs`, `light`, `abci`, `crypto` and `types` (or whatever succeeds types) in the public space? This vastly reduces the public surface area, and is less impactful to the code, at the expense of making tendermint usable (or ergonomic) as a library. Users won't be able to inject custom reactors or implementations and the experience of using tendermint as a library will be roughly the same as using it as a standalone binary, with all configuration via the configuration structures. This might increase the likelihood that library users will fork to use tendermint as a library. 

- minimalistically we could avoid an internal package entirely, and make `Reactor` and interface and put public constructors that return the public interfaces (rather than the concrete types as is the case now) in all packages and make *all* of the existing types and implementations. The interfaces and constructors for the reactors would need to be public and exported, but the implementations could be entirely private. This is probably more labor intensive to do, but means that it will be possible to inject alternate implementations and construct tendermint instances with substantially different behavior. This will likely create a larger API surface area, but might make library use without-forking more common, at the risk that supporting some of these dependency injection use-cases may be more delicate.

- theoretically there's a moderate position between these, where the interfaces for the components (reactors, mempools, possibly state machines,) are either directly within the `tendermint` root, and most of the implementations and constructors stored separately in some kind of internal package. This would keep some porting (say) `node` public, and would allow some injection but not the use of `tendermint` own implementation. The dependency and constructability requirements might make this substantially worse than other options.

Conclusions: 
- if we believe in letting users of tendermint's APIs compose their own services, and run tendermint with their own components, then we should do the minimalist approach. 
- if we believe that "library" users and "binary" should have the same experience, then we should do the maximaslist approach.

**Comment by tac0turtle** [comment]
Nice writeup, have you had a chance to read over the ADR that was written by erik (https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-060-go-api-stability.md)?

**Reply by tychoish** [reply]
Yeah, I have, and I think the core goals of that document are definitely good and achievable (e.g. "stabilize the API for 1.0", "use semver", "reduce the API surface area by moving things to internal"). I think there's a bit of ambiguity in terms of where exactly the line is (and the `node` package makes this a bit thorny, as I've begun to pull on the strings,) and I think the ADR (or this thread) would be greatly illuminated/improved by crystallizing the user stories around library use.
